runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/solver/pathfinding/IDAStarMM.java

```diff
@@ -0,0 +1,787 @@
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.puzzlesolver.solver.pathfinding;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import net.runelite.api.Point;
+import static net.runelite.client.plugins.puzzlesolver.solver.PuzzleSolver.BLANK_TILE_VALUE;
+import static net.runelite.client.plugins.puzzlesolver.solver.PuzzleSolver.DIMENSION;
+import net.runelite.client.plugins.puzzlesolver.solver.PuzzleState;
+import net.runelite.client.plugins.puzzlesolver.solver.heuristics.Heuristic;
+
+public class IDAStarMM extends IDAStar
+{
+	private PuzzleState currentState;
+	private List<PuzzleState> stateList = new ArrayList<>();
+	private List<List<Integer>> validRowNumbers = new ArrayList<>();
+	private List<List<Integer>> validColumnNumbers = new ArrayList<>();
+
+	public IDAStarMM(Heuristic heuristic)
+	{
+		super(heuristic);
+
+		//Add valid numbers for rows and columns
+		validRowNumbers.add(Arrays.asList(0, 1, 2, 3, 4));
+		validRowNumbers.add(Arrays.asList(6, 7, 8, 9));
+		validColumnNumbers.add(Arrays.asList(5, 10, 15, 20));
+	}
+
+	@Override
+	public List<PuzzleState> computePath(PuzzleState root)
+	{
+		currentState = root;
+		stateList.add(root);
+
+		List<PuzzleState> path = new ArrayList<>();
+
+		//Reduce to 4x5
+		solveRow(0);
+
+		//Reduce to 4x4
+		solveColumn();
+
+		//Reduce to 3x4
+		solveRow(1);
+
+		//Remove last state
+		stateList.remove(stateList.size() - 1);
+
+		//Pathfinder for 4x4
+		path.addAll(super.computePath(currentState));
+
+		path.addAll(0, stateList);
+
+		return path;
+	}
+
+	private void solveRow(int row)
+	{
+		for (int i = row; i < DIMENSION; i++)
+		{
+			int valTarget = row * DIMENSION + i;
+
+			int valCurrent = currentState.getPiece(i, row);
+
+			if (valCurrent != valTarget)
+			{
+				moveTowardsVal(valTarget, i, row,  true);
+			}
+		}
+	}
+
+	private void solveColumn()
+	{
+		int column = 0;
+
+		for (int i = column + 1; i < DIMENSION; i++)
+		{
+			int valTarget = column + i * DIMENSION;
+
+			int valCurrent = currentState.getPiece(column, i);
+
+			if (valCurrent != valTarget)
+			{
+				moveTowardsVal(valTarget, column, i,  false);
+			}
+		}
+	}
+
+	private void moveTowardsVal(int valTarget, int x, int y, boolean rowMode)
+	{
+		//Not in place
+		boolean reached = false;
+
+		int row = y;
+		int column = x;
+
+		while (currentState.getPiece(x, y) != valTarget)
+		{
+			//Find piece location
+			Point locVal = findPiece(valTarget);
+			Point locBlank = findPiece(BLANK_TILE_VALUE);
+
+			if (!reached)
```

Comment:
Consider flipping this condition so the small code block comes first, then ``else``, and then the large block.